File: python/client/convex.py
"""A deliberately small, native Python Convex demonstration client."""
import base64
import hashlib
import json
import os
import queue
import secrets
import socket
import ssl
import struct
import threading
import time
import urllib.error
import urllib.parse
import urllib.request
from dataclasses import dataclass


class ConvexError(Exception): pass
class ClosedError(ConvexError): pass
class ProtocolError(ConvexError): pass
class TransportError(ConvexError): pass


class FunctionError(ConvexError):
    def __init__(self, message, *, data=None, logs=None):
        super().__init__(message); self.data = data; self.logs = logs or []


@dataclass
class Result:
    value: object
    logs: list


@dataclass
class Update:
    value: object = None
    error: Exception | None = None
    logs: list | None = None


class _WebSocket:
    # RFC 6455 framing is kept here instead of hiding Convex behaviour behind a
    # third-party SDK. It supports the text frames used by the pinned profile.
    def __init__(self, url, client_version):
        parsed = urllib.parse.urlparse(url)
        raw = socket.create_connection((parsed.hostname, parsed.port or (443 if parsed.scheme == "wss" else 80)), 10)
        self.sock = ssl.create_default_context().wrap_socket(raw, server_hostname=parsed.hostname) if parsed.scheme == "wss" else raw
        key = base64.b64encode(os.urandom(16)).decode()
        path = (parsed.path or "/") + (("?" + parsed.query) if parsed.query else "")
        self.sock.sendall((f"GET {path} HTTP/1.1\r\nHost: {parsed.netloc}\r\nUpgrade: websocket\r\nConnection: Upgrade\r\nSec-WebSocket-Key: {key}\r\nSec-WebSocket-Version: 13\r\nConvex-Client: {client_version}\r\n\r\n").encode())
        response = self._read_until(b"\r\n\r\n")
        if not response.startswith(b"HTTP/1.1 101") or base64.b64encode(hashlib.sha1((key + "258EAFA5-E914-47DA-95CA-C5AB0DC85B11").encode()).digest()) not in response:
            self.close(); raise TransportError("WebSocket upgrade was rejected")

    def _read_until(self, marker):
        data = b""
        while marker not in data:
            chunk = self.sock.recv(4096)
            if not chunk: raise TransportError("WebSocket closed during upgrade")
            data += chunk
        return data

    def _exact(self, length):
        data = b""
        while len(data) < length:
            chunk = self.sock.recv(length - len(data))
            if not chunk: raise EOFError
            data += chunk
        return data

    def send_json(self, value):
        payload = json.dumps(value, separators=(",", ":")).encode(); mask = os.urandom(4)
        size = len(payload)
        head = bytes([0x81, 0x80 | (size if size < 126 else 126 if size < 65536 else 127)])
        if size >= 65536: head += struct.pack("!Q", size)
        elif size >= 126: head += struct.pack("!H", size)
        masked = bytes(byte ^ mask[index % 4] for index, byte in enumerate(payload))
        self.sock.sendall(head + mask + masked)

    def receive_json(self):
        first, second = self._exact(2); opcode = first & 15; size = second & 127
        if size == 126: size = struct.unpack("!H", self._exact(2))[0]
        elif size == 127: size = struct.unpack("!Q", self._exact(8))[0]
        mask = self._exact(4) if second & 128 else None; data = self._exact(size)
        if mask: data = bytes(byte ^ mask[index % 4] for index, byte in enumerate(data))
        if opcode == 8: raise EOFError
        if opcode == 9: self.sock.sendall(b"\x8a\x00"); return self.receive_json()
        if opcode != 1: raise ProtocolError(f"unsupported WebSocket opcode {opcode}")
        return json.loads(data)

    def close(self):
        try: self.sock.close()
        except Exception: pass


class Subscription:
    MAX_BUFFERED_UPDATES = 16
    def __init__(self, manager, query_id): self.manager, self.query_id, self.updates, self.closed = manager, query_id, queue.Queue(self.MAX_BUFFERED_UPDATES), False
    def next_update(self, timeout=None):
        try: update = self.updates.get(timeout=timeout)
        except queue.Empty: raise TransportError("timed out waiting for Live update")
        if isinstance(update, ClosedError): raise update
        return update
    def deliver(self, update):
        if not self.closed:
            try: self.updates.put_nowait(update)
            except queue.Full:
                try: self.updates.get_nowait()
                except queue.Empty: pass
                self.updates.put_nowait(update)
    def close(self):
        if not self.closed: self.closed = True; self.manager.unsubscribe(self.query_id)


class _LiveManager:
    INITIAL_VERSION = {"querySet": 0, "identity": 0, "ts": "AAAAAAAAAAA="}
    def __init__(self, url, version):
        p = urllib.parse.urlparse(url); self.url = urllib.parse.urlunparse(("wss" if p.scheme == "https" else "ws", p.netloc, p.path.rstrip("/") + "/api/sync", "", "", "")); self.version = version
        self.subs, self.socket, self.lock, self.closed, self.connection_count = {}, None, threading.RLock(), False, 0
        self.thread = threading.Thread(target=self._run, daemon=True); self.thread.start()
    def subscribe(self, path, args):
        with self.lock:
            ident = max(self.subs, default=-1) + 1; sub = Subscription(self, ident); self.subs[ident] = (path, args, sub)
            if self.socket: self._modify([self._add(ident, path, args)])
            return sub
    def unsubscribe(self, ident):
        with self.lock:
            if self.subs.pop(ident, None) and self.socket: self._modify([{ "type": "Remove", "queryId": ident }])
    def _add(self, ident, path, args): return {"type":"Add", "queryId":ident, "udfPath":path, "args":[args]}
    def _connect(self):
        self.socket = _WebSocket(self.url, self.version); self.query_version = 0; self.remote_version = dict(self.INITIAL_VERSION)
        self.socket.send_json({"type":"Connect", "sessionId":secrets.token_hex(16), "connectionCount":self.connection_count, "lastCloseReason":"InitialConnect", "clientTs":0})
        self._modify([self._add(i, p, a) for i, (p,a,_) in self.subs.items()])
    def _modify(self, mods):
        if not mods: return
        self.socket.send_json({"type":"ModifyQuerySet", "baseVersion":self.query_version, "newVersion":self.query_version + 1, "modifications":mods}); self.query_version += 1
    def _run(self):
        while not self.closed:
            try:
                with self.lock:
                    if not self.subs: time.sleep(.05); continue
                    if not self.socket: self._connect()
                    message = self.socket.receive_json()
                if message.get("type") == "Transition": self._transition(message)
                elif message.get("type") in ("Ping", "MutationResponse", "ActionResponse"): pass
                elif message.get("type") == "TransitionChunk": raise ProtocolError("TransitionChunk assembly is deferred")
                else: raise ProtocolError(f"unexpected Live message {message.get('type')!r}")
            except Exception as error:
                with self.lock:
                    if self.socket: self.socket.close(); self.socket = None; self.connection_count += 1
                    for _,_,sub in self.subs.values(): sub.deliver(Update(error=error, logs=[]))
                time.sleep(.1)
    def _transition(self, message):
        if message.get("startVersion") != self.remote_version: raise ProtocolError("Transition start version does not match local version")
        changed=[]
        for mod in message.get("modifications", []):
            ident=mod["queryId"]
            if mod["type"] == "QueryUpdated": changed.append((ident, Update(mod.get("value"), logs=mod.get("logLines", []))))
            elif mod["type"] == "QueryFailed": changed.append((ident, Update(error=FunctionError(mod.get("errorMessage", "query failed"), data=mod.get("errorData"), logs=mod.get("logLines", [])), logs=mod.get("logLines", []))))
            elif mod["type"] != "QueryRemoved": raise ProtocolError(f"unknown Transition modification {mod['type']!r}")
        self.remote_version=message["endVersion"]
        with self.lock:
            for ident,update in changed:
                if ident in self.subs: self.subs[ident][2].deliver(update)
    def disconnect_for_adapter(self):
        with self.lock:
            if not self.socket: raise TransportError("Live WebSocket is not connected")
            self.socket.close(); self.socket=None
    def close(self):
        self.closed=True
        with self.lock:
            if self.socket: self.socket.close()
            for _,_,sub in self.subs.values(): sub.deliver(ClosedError("Live subscription is closed"))


class Client:
    VERSION = "python-0.1.0"
    def __init__(self, deployment_url, bearer_token=None):
        p=urllib.parse.urlparse(deployment_url)
        if p.scheme not in ("http", "https") or not p.hostname or p.username: raise ValueError("Convex deployment URL must be an http(s) URL with a host")
        self.url=deployment_url.rstrip("/"); self.token=bearer_token or ""; self.closed=False; self.live=None
    def set_auth(self, token): self._check(); self.token=token or ""
    def _check(self):
        if self.closed: raise ClosedError("Convex client is closed")
    def _call(self, operation, path, args):
        self._check()
        if not isinstance(args, dict) or not path: raise ValueError("Convex path and named object arguments are required")
        request=urllib.request.Request(self.url + "/api/" + operation, data=json.dumps({"path":path,"args":args,"format":"json"}).encode(), headers={"Content-Type":"application/json", "Accept":"application/json", "Convex-Client":self.VERSION, **({"Authorization":"Bearer " + self.token} if self.token else {})}, method="POST")
        try:
            with urllib.request.urlopen(request, timeout=30) as response: decoded=json.load(response)
        except (urllib.error.URLError, OSError, json.JSONDecodeError) as error: raise TransportError(str(error)) from error
        if decoded.get("status") == "success" and "value" in decoded: return Result(decoded["value"], decoded.get("logLines", []))
        if decoded.get("status") == "error": raise FunctionError(decoded.get("errorMessage", "Convex function failed"), data=decoded.get("errorData"), logs=decoded.get("logLines", []))
        raise ProtocolError("unknown Convex HTTP response")
    def query(self,path,args=None): return self._call("query",path,args or {})
    def mutation(self,path,args=None): return self._call("mutation",path,args or {})
    def action(self,path,args=None): return self._call("action",path,args or {})
    def subscribe(self,path,args=None):
        self._check(); self.live=self.live or _LiveManager(self.url,self.VERSION); return self.live.subscribe(path,args or {})
    def debug_disconnect_for_adapter(self): self._check(); self.live.disconnect_for_adapter() if self.live else (_ for _ in ()).throw(TransportError("Live WebSocket is not connected"))
    def close(self):
        if not self.closed: self.closed=True; self.live and self.live.close()

File: php/client/convex.php
<?php
declare(strict_types=1);

/* Native, dependency-free Convex experiment.  HTTP is documented; the small
 * RFC6455 transport below is deliberately kept separate from the pinned sync
 * state machine so readers can see which part is protocol-specific. */
namespace Convex;

class Error extends \RuntimeException {}
class ProtocolError extends Error {}
class ClosedError extends Error {}
class TransportError extends Error
{
    public function __construct(
        string $m,
        public string $operation = "transport",
    ) {
        parent::__construct("Convex $operation transport error: $m");
    }
}
class FunctionError extends Error
{
    public function __construct(
        string $m,
        public string $operation,
        public mixed $data = null,
        public array $logs = [],
    ) {
        parent::__construct("Convex $operation failed: $m");
    }
}
final class Result
{
    public function __construct(public mixed $value, public array $logs = []) {}
}
final class Update
{
    public function __construct(
        public mixed $value = null,
        public ?Error $error = null,
        public array $logs = [],
    ) {}
}

final class Client
{
    private string $url;
    private string $token = "";
    private bool $closed = false;
    private ?LiveManager $live = null;
    public function __construct(
        string $url,
        ?string $bearerToken = null,
        private string $version = "php-0.1.0",
    ) {
        $p = parse_url($url);
        if (
            !is_array($p) ||
            !isset($p["scheme"], $p["host"]) ||
            !in_array($p["scheme"], ["http", "https"], true) ||
            isset($p["user"])
        ) {
            throw new \InvalidArgumentException(
                "Convex deployment URL must be http(s) with a host",
            );
        }
        $this->url = rtrim($url, "/");
        $this->token = (string) $bearerToken;
    }
    public function setAuth(string $token): void
    {
        $this->guard();
        $this->token = $token;
    }
    public function query(string $path, array $args = []): Result
    {
        return $this->call("query", $path, $args);
    }
    public function mutation(string $path, array $args = []): Result
    {
        return $this->call("mutation", $path, $args);
    }
    public function action(string $path, array $args = []): Result
    {
        return $this->call("action", $path, $args);
    }
    public function subscribe(string $path, array $args = []): Subscription
    {
        $this->validate($path, $args);
        $this->guard();
        return ($this->live ??= new LiveManager(
            $this->url,
            $this->version,
        ))->subscribe($path, $args);
    }
    public function pump(float $timeout = 0.0): void
    {
        $this->live?->pump($timeout);
    }
    public function debugDisconnectForAdapter(): void
    {
        $this->guard();
        if (!$this->live) {
            throw new TransportError("Live WebSocket is not connected", "live");
        }
        $this->live->debugDisconnect();
    }
    public function close(): void
    {
        if (!$this->closed) {
            $this->closed = true;
            $this->live?->close();
        }
    }
    private function guard(): void
    {
        if ($this->closed) {
            throw new ClosedError("Convex client is closed");
        }
    }
    private function validate(string $path, array $args): void
    {
        if ($path === "") {
            throw new \InvalidArgumentException(
                "Convex function path is required",
            );
        }
        try {
            json_encode(
                $args,
                JSON_THROW_ON_ERROR | JSON_INVALID_UTF8_SUBSTITUTE,
            );
        } catch (\JsonException $e) {
            throw new \InvalidArgumentException(
                "encode Convex arguments: " . $e->getMessage(),
            );
        }
    }
    private function call(string $op, string $path, array $args): Result
    {
        $this->guard();
        $this->validate($path, $args);
        $body = json_encode(
            ["path" => $path, "args" => $args, "format" => "json"],
            JSON_THROW_ON_ERROR | JSON_INVALID_UTF8_SUBSTITUTE,
        );
        $headers =
            "Content-Type: application/json\r\nAccept: application/json\r\nConvex-Client: {$this->version}\r\n" .
            ($this->token === ""
                ? ""
                : "Authorization: Bearer {$this->token}\r\n");
        $ctx = stream_context_create([
            "http" => [
                "method" => "POST",
                "header" => $headers,
                "content" => $body,
                "timeout" => 30,
                "ignore_errors" => true,
            ],
        ]);
        $raw = @file_get_contents("{$this->url}/api/$op", false, $ctx);
        if ($raw === false) {
            throw new TransportError(
                error_get_last()["message"] ?? "HTTP request failed",
                $op,
            );
        }
        if (strlen($raw) > 2097152) {
            throw new TransportError("response exceeds 2097152 bytes", $op);
        }
        try {
            $decoded = json_decode($raw, true, 512, JSON_THROW_ON_ERROR);
        } catch (\JsonException $e) {
            throw new TransportError(
                "HTTP response was not JSON: " . $e->getMessage(),
                $op,
            );
        }
        if (
            ($decoded["status"] ?? null) === "success" &&
            array_key_exists("value", $decoded)
        ) {
            return new Result(
                $decoded["value"],
                array_values($decoded["logLines"] ?? []),
            );
        }
        if (($decoded["status"] ?? null) === "error") {
            throw new FunctionError(
                (string) ($decoded["errorMessage"] ?? ""),
                $op,
                $decoded["errorData"] ?? null,
                array_values($decoded["logLines"] ?? []),
            );
        }
        throw new ProtocolError("HTTP response has unknown Convex status");
    }
}

final class Subscription
{
    /** newest sixteen wins: slow consumers retain the latest reactive state. */
    private array $updates = [];
    private bool $finished = false;
    public function __construct(private LiveManager $manager, public int $id) {}
    public function nextUpdate(float $timeout = 10): Update
    {
        $end = microtime(true) + $timeout;
        while (!$this->updates && !$this->finished) {
            $this->manager->pump(max(0.0, min(0.1, $end - microtime(true))));
            if (microtime(true) >= $end) {
                throw new TransportError(
                    "timed out waiting for Live update",
                    "live",
                );
            }
        }
        if (!$this->updates) {
            throw new ClosedError("Live subscription is closed");
        }
        return array_shift($this->updates);
    }
    public function deliver(Update $u): void
    {
        if (!$this->finished) {
            if (count($this->updates) === 16) {
                array_shift($this->updates);
            }
            $this->updates[] = $u;
        }
    }
    public function close(): void
    {
        if (!$this->finished) {
            $this->manager->unsubscribe($this->id);
        }
    }
    public function finish(): void
    {
        $this->finished = true;
        $this->updates = [];
    }
}

final class LiveManager
{
    private ?WebSocket $ws = null;
    private array $subs = [];
    private int $next = 0,
        $querySet = 0,
        $connections = 0;
    private array $remote = [
        "querySet" => 0,
        "identity" => 0,
        "ts" => "AAAAAAAAAAA=",
    ];
    private float $reconnectAt = 0,
        $backoff = 0.1;
    private string $lastClose = "InitialConnect";
    private bool $closed = false;
    public function __construct(private string $url, private string $version) {}
    public function subscribe(string $path, array $args): Subscription
    {
        if ($this->closed) {
            throw new ClosedError("Live manager is closed");
        }
        $s = new Subscription($this, $this->next++);
        $this->subs[$s->id] = [
            "path" => $path,
            "args" => $args,
            "subscription" => $s,
        ];
        $this->ensure();
        if ($this->ws) {
            $this->modify([self::add($s->id, $this->subs[$s->id])]);
        }
        return $s;
    }
    public function unsubscribe(int $id): void
    {
        $state = $this->subs[$id] ?? null;
        unset($this->subs[$id]);
        if ($state) {
            $state["subscription"]->finish();
            if ($this->ws) {
                $this->modify([["type" => "Remove", "queryId" => $id]]);
            }
        }
    }
    public function debugDisconnect(): void
    {
        if (!$this->ws) {
            $this->ensure();
            if (!$this->ws) {
                throw new TransportError(
                    "Live WebSocket is not connected",
                    "live",
                );
            }
        }
        // Adapter-only: confirm the old transport is closed, then return so the
        // adapter can acknowledge before an external mutation is made. The normal
        // 100 ms reconnect path will resubscribe and observe that newer value.
        $this->disconnect("DebugDisconnect");
    }
    public function close(): void
    {
        $this->closed = true;
        $this->ws?->close();
        $this->ws = null;
        foreach ($this->subs as $s) {
            $s["subscription"]->finish();
        }
        $this->subs = [];
    }
    public function pump(float $timeout = 0): void
    {
        if ($this->closed) {
            return;
        }
        $this->ensure();
        if (!$this->ws) {
            if ($timeout > 0) {
                usleep((int) ($timeout * 1000000));
            }
            return;
        }
        $this->ws->wait($timeout);
        while (($raw = $this->ws->read()) !== false) {
            if ($raw === null) {
                $this->disconnect("ServerClosed");
                break;
            }
            $this->handle($raw);
        }
    }
    private function ensure(): void
    {
        if ($this->ws || !$this->subs || microtime(true) < $this->reconnectAt) {
            return;
        }
        try {
            $p = parse_url($this->url);
            $scheme = $p["scheme"] === "https" ? "wss" : "ws";
            $host = $p["host"];
            $port = isset($p["port"]) ? ":" . $p["port"] : "";
            $path = rtrim($p["path"] ?? "", "/") . "/api/sync";
            $this->ws = new WebSocket(
                "$scheme://$host$port$path",
                $this->version,
            );
            $this->querySet = 0;
            $this->remote = [
                "querySet" => 0,
                "identity" => 0,
                "ts" => "AAAAAAAAAAA=",
            ];
            $this->ws->send([
                "type" => "Connect",
                "sessionId" => self::uuid(),
                "connectionCount" => $this->connections,
                "lastCloseReason" => $this->lastClose,
                "clientTs" => 0,
            ]);
            $mods = [];
            foreach ($this->subs as $id => $s) {
                $mods[] = self::add($id, $s);
            }
            if ($mods) {
                $this->modify($mods);
            }
            $this->backoff = 0.1;
        } catch (\Throwable $e) {
            $this->disconnect($e->getMessage());
        }
    }
    private function modify(array $mods): void
    {
        if (!$this->ws) {
            return;
        }
        $this->ws->send([
            "type" => "ModifyQuerySet",
            "baseVersion" => $this->querySet,
            "newVersion" => $this->querySet + 1,
            "modifications" => $mods,
        ]);
        $this->querySet++;
    }
    private static function add(int $id, array $s): array
    {
        return [
            "type" => "Add",
            "queryId" => $id,
            "udfPath" => $s["path"],
            "args" => [$s["args"]],
        ];
    }
    private function handle(string $raw): void
    {
        try {
            $m = json_decode($raw, true, 512, JSON_THROW_ON_ERROR);
        } catch (\JsonException $e) {
            throw new ProtocolError(
                "decode server message: " . $e->getMessage(),
            );
        }
        if (($m["type"] ?? "") !== "Transition") {
            if (
                in_array(
                    $m["type"] ?? "",
                    ["Ping", "MutationResponse", "ActionResponse"],
                    true,
                )
            ) {
                return;
            }
            throw new ProtocolError(
                "unknown server message " . ($m["type"] ?? ""),
            );
        }
        if (($m["startVersion"] ?? null) != $this->remote) {
            throw new ProtocolError("Transition start version mismatch");
        }
        $changed = [];
        foreach ($m["modifications"] ?? [] as $x) {
            $id = $x["queryId"];
            if (($x["type"] ?? "") === "QueryUpdated") {
                $changed[$id] = new Update(
                    $x["value"] ?? null,
                    null,
                    array_values($x["logLines"] ?? []),
                );
            } elseif (($x["type"] ?? "") === "QueryFailed") {
                $changed[$id] = new Update(
                    null,
                    new FunctionError(
                        (string) ($x["errorMessage"] ?? ""),
                        "query",
                        $x["errorData"] ?? null,
                        array_values($x["logLines"] ?? []),
                    ),
                    array_values($x["logLines"] ?? []),
                );
            }
        }
        $this->remote = $m["endVersion"];
        foreach ($changed as $id => $u) {
            if (isset($this->subs[$id])) {
                $this->subs[$id]["subscription"]->deliver($u);
            }
        }
    }
    private function disconnect(string $reason): void
    {
        $this->ws?->close();
        $this->ws = null;
        $this->connections++;
        $this->lastClose = $reason;
        $this->querySet = 0;
        $this->remote = [
            "querySet" => 0,
            "identity" => 0,
            "ts" => "AAAAAAAAAAA=",
        ];
        $this->reconnectAt = microtime(true) + $this->backoff;
        $this->backoff = min(15, $this->backoff * 2);
    }
    private static function uuid(): string
    {
        $b = random_bytes(16);
        $b[6] = chr((ord($b[6]) & 15) | 64);
        $b[8] = chr((ord($b[8]) & 63) | 128);
        return vsprintf("%s%s-%s-%s-%s-%s%s%s", str_split(bin2hex($b), 4));
    }
}

final class WebSocket
{
    private $io;
    private string $buffer = "";
    private string $fragments = "";
    private ?int $fragmentOpcode = null;
    public function __construct(string $url, string $version)
    {
        $p = parse_url($url);
        $ssl = $p["scheme"] === "wss";
        $target =
            ($ssl ? "ssl" : "tcp") .
            "://" .
            $p["host"] .
            ":" .
            ($p["port"] ?? ($ssl ? 443 : 80));
        $this->io = @stream_socket_client(
            $target,
            $e,
            $s,
            10,
            STREAM_CLIENT_CONNECT,
            stream_context_create([
                "ssl" => ["verify_peer" => true, "peer_name" => $p["host"]],
            ]),
        );
        if (!$this->io) {
            throw new TransportError("WebSocket connect: $s", "live");
        }
        stream_set_blocking($this->io, true);
        $key = base64_encode(random_bytes(16));
        $resource =
            ($p["path"] ?? "/") . (isset($p["query"]) ? "?" . $p["query"] : "");
        fwrite(
            $this->io,
            "GET $resource HTTP/1.1\r\nHost: {$p["host"]}\r\nUpgrade: websocket\r\nConnection: Upgrade\r\nSec-WebSocket-Key: $key\r\nSec-WebSocket-Version: 13\r\nConvex-Client: $version\r\n\r\n",
        );
        $h = "";
        while (!str_contains($h, "\r\n\r\n")) {
            $c = fread($this->io, 1);
            if ($c === "" || $c === false) {
                throw new ProtocolError("WebSocket upgrade closed");
            }
            $h .= $c;
            if (strlen($h) > 32768) {
                throw new ProtocolError("WebSocket headers too large");
            }
        }
        $lines = explode("\r\n", $h);
        if (!str_contains($lines[0], " 101 ")) {
            throw new ProtocolError("WebSocket upgrade failed");
        }
        $headers = [];
        foreach ($lines as $l) {
            if (str_contains($l, ":")) {
                [$a, $b] = explode(":", $l, 2);
                $headers[strtolower($a)] = trim($b);
            }
        }
        if (
            ($headers["sec-websocket-accept"] ?? "") !==
            base64_encode(
                sha1($key . "258EAFA5-E914-47DA-95CA-C5AB0DC85B11", true),
            )
        ) {
            throw new ProtocolError("invalid WebSocket accept");
        }
        stream_set_blocking($this->io, false);
    }
    public function wait(float $seconds): void
    {
        $r = [$this->io];
        $w = $e = [];
        @stream_select(
            $r,
            $w,
            $e,
            (int) $seconds,
            (int) (($seconds - (int) $seconds) * 1000000),
        );
    }
    public function send(array $v): void
    {
        $this->frame(
            1,
            json_encode($v, JSON_THROW_ON_ERROR | JSON_INVALID_UTF8_SUBSTITUTE),
        );
    }
    public function read(): string|false|null
    {
        $chunk = stream_get_contents($this->io);
        $this->buffer .= $chunk ?: "";
        if ($this->buffer === "" && feof($this->io)) {
            return null;
        }
        while (true) {
            if (strlen($this->buffer) < 2) {
                return false;
            }
            [$a, $b] = array_values(unpack("C2", substr($this->buffer, 0, 2)));
            $fin = (bool) ($a & 128);
            $len = $b & 127;
            $at = 2;
            if ($b & 128) {
                throw new ProtocolError(
                    "server WebSocket frames must not be masked",
                );
            }
            if ($len === 126) {
                if (strlen($this->buffer) < 4) {
                    return false;
                }
                $len = unpack("n", substr($this->buffer, 2, 2))[1];
                $at = 4;
            } elseif ($len === 127) {
                throw new ProtocolError("oversized WebSocket frame");
            }
            if ($len > 2097152) {
                throw new ProtocolError("WebSocket frame too large");
            }
            if (strlen($this->buffer) < $at + $len) {
                return false;
            }
            $payload = substr($this->buffer, $at, $len);
            $this->buffer = substr($this->buffer, $at + $len);
            $op = $a & 15;
            if ($op === 8) {
                return null;
            }
            if ($op === 9) {
                $this->frame(10, $payload);
                continue;
            }
            if ($op === 10) {
                continue;
            }
            if ($op === 1) {
                if ($this->fragmentOpcode !== null) {
                    throw new ProtocolError("interleaved WebSocket text");
                }
                $this->fragments = $payload;
                $this->fragmentOpcode = 1;
            } elseif ($op === 0) {
                if ($this->fragmentOpcode === null) {
                    throw new ProtocolError(
                        "unexpected WebSocket continuation",
                    );
                }
                $this->fragments .= $payload;
            } else {
                throw new ProtocolError("unsupported WebSocket frame");
            }
            if (strlen($this->fragments) > 2097152) {
                throw new ProtocolError("WebSocket message too large");
            }
            if (!$fin) {
                continue;
            }
            $message = $this->fragments;
            $this->fragments = "";
            $this->fragmentOpcode = null;
            return $message;
        }
    }
    private function frame(int $op, string $payload): void
    {
        $n = strlen($payload);
        $mask = random_bytes(4);
        $head =
            chr(128 | $op) .
            ($n < 126 ? chr(128 | $n) : chr(254) . pack("n", $n));
        $out = "";
        for ($i = 0; $i < $n; $i++) {
            $out .= $payload[$i] ^ $mask[$i % 4];
        }
        fwrite($this->io, $head . $mask . $out);
    }
    public function close(): void
    {
        if (is_resource($this->io)) {
            fclose($this->io);
        }
    }
}
